### ISSUE TITLE
Pointer artithmetic

### ISSUE THREAD

**Issue by CToups** [johnwfinigan/CSCI2467-Fall2015#13]
Seems to me pointer subtraction is just regular subtraction of the indexes pointed to. The C11 standard says:
"In other words, if the expressions P and Q point to, respectively, the i-th and j-th elements of
an array object, the expression (P)-(Q) has the value i−j provided the value fits in an object of type ptrdiff_t."

Which sounds like regular subtraction? What am I missing?

**Comment by johnwfinigan** [comment]
You've basically got it. The key is what **type** is on each side of the `-` operator. Roughly:

integer `-` integer yields integer   (in the normal arithmetic sense)

pointer `-` integer yields pointer. Standard says: 

`When an expression that has integer type is added to or subtracted from a pointer, the
result has the type of the pointer operand. If the pointer operand points to an element of
an array object, and the array is large enough, the result points to an element offset from
the original element such that the difference of the subscripts of the resulting and original
array elements equals the integer expression.`

In other words, if `x` points to `a[3]`,  `x - 2` points to `a[1]`

pointer `-` pointer yields integer. Standard says:

`When two pointers are subtracted, both shall point to elements of the same array object,
or one past the last element of the array object; the result is the difference of the
subscripts of the two array elements.`

In other words, if `x` points to `a[3]` and `y` points to `a[5]`, `y - x`  yields 2.

Helps?

**Comment by johnwfinigan** [comment]
Standard citations are from page 93 of the pdf below, section 6.5.6

http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf

**Comment by johnwfinigan** [comment]
Note, last line of my first response was edited, originally said `x - y`, correctly says `y - x`

**Comment by CToups** [comment]
Very helpful thank you
